File: sdk/javascript/src/webhooks.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import type { WebhookEvent } from "./types";
import { AuthenticationError } from "./errors";

/**
 * Verify an Oelala webhook signature (HMAC-SHA256).
 *
 * @param payload - Raw request body (string or Buffer)
 * @param signature - The `X-Webhook-Signature` header value (format: `sha256=<hex>`)
 * @param secret - Your webhook secret (format: `whsec_...`)
 * @returns `true` if valid
 * @throws {AuthenticationError} If the signature is invalid
 *
 * @example
 * ```ts
 * import { verifyWebhookSignature } from "@oelala/sdk";
 *
 * // Express.js example
 * app.post("/webhook", express.raw({ type: "application/json" }), (req, res) => {
 *   verifyWebhookSignature(
 *     req.body,
 *     req.headers["x-webhook-signature"],
 *     process.env.WEBHOOK_SECRET
 *   );
 *   const event = JSON.parse(req.body);
 *   console.log("Event:", event.event, event.data.job_id);
 *   res.sendStatus(200);
 * });
 * ```
 */
export function verifyWebhookSignature(
  payload: string | Buffer,
  signature: string,
  secret: string
): boolean {
  if (!signature.startsWith("sha256=")) {
    throw new AuthenticationError("Invalid signature format: must start with 'sha256='");
  }

  const expectedHex = signature.slice(7); // strip "sha256="
  const computed = createHmac("sha256", secret)
    .update(typeof payload === "string" ? payload : payload)
    .digest("hex");

  const a = Buffer.from(computed, "hex");
  const b = Buffer.from(expectedHex, "hex");

  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    throw new AuthenticationError("Invalid webhook signature");
  }

  return true;
}

/**
 * Parse a webhook payload into a typed WebhookEvent.
 *
 * @param payload - Parsed JSON body
 */
export function parseWebhookEvent(payload: Record<string, unknown>): WebhookEvent {
  return {
    event: payload.event as string,
    event_id: (payload.event_id as string) ?? "",
    timestamp: (payload.timestamp as string) ?? "",
    data: (payload.data as Record<string, unknown>) ?? {},
  };
}

File: sdk/python/oelala/webhooks.py
"""Webhook signature verification utilities."""

from __future__ import annotations

import hashlib
import hmac
from typing import Union

from oelala.exceptions import AuthenticationError
from oelala.models import WebhookEvent


def verify_webhook_signature(
    payload: Union[str, bytes],
    signature: str,
    secret: str,
) -> bool:
    """Verify an Oelala webhook signature.

    Args:
        payload: The raw request body (string or bytes).
        signature: The ``X-Webhook-Signature`` header value
            (format: ``sha256=<hex>``).
        secret: Your webhook secret (format: ``whsec_...``).

    Returns:
        True if the signature is valid.

    Raises:
        AuthenticationError: If the signature is invalid or malformed.

    Example::

        from oelala import verify_webhook_signature

        # In your webhook handler (e.g. Flask, FastAPI, Django)
        is_valid = verify_webhook_signature(
            payload=request.body,
            signature=request.headers["X-Webhook-Signature"],
            secret="whsec_your_secret",
        )
    """
    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    if not signature.startswith("sha256="):
        raise AuthenticationError("Invalid signature format: must start with 'sha256='")

    expected_hex = signature[7:]  # strip "sha256="

    computed = hmac.new(
        secret.encode("utf-8"),
        payload,
        hashlib.sha256,
    ).hexdigest()

    if not hmac.compare_digest(computed, expected_hex):
        raise AuthenticationError("Invalid webhook signature")

    return True


def parse_webhook_event(payload: dict) -> WebhookEvent:
    """Parse a webhook payload dict into a WebhookEvent.

    Args:
        payload: The parsed JSON body of the webhook request.

    Returns:
        :class:`WebhookEvent` instance.
    """
    return WebhookEvent(
        event=payload["event"],
        event_id=payload.get("event_id", ""),
        timestamp=payload.get("timestamp", ""),
        data=payload.get("data", {}),
    )
